slurm_openstack_tools/reboot.py

```diff
@@ -1,160 +1,122 @@
-            "--noheader",
-            "--nodes=%s" % hostname,
-            "-O",
-            "Reason:%i" % MAX_REASON_LENGTH,
-        ],
-        stdout=subprocess.PIPE,
-        universal_newlines=True,
-    )
-    return sinfo.stdout.strip()
-
-
-def get_image_from_reason(reason):
-    tokens = reason.split()
-    image = None
-    if len(tokens) > 1:
-        image_tokens = tokens[1].split(":")
-        if len(image_tokens) == 2 and image_tokens[0] == "image":
-            if image_tokens[1]:
-                image = image_tokens[1]
-                logger.info(f"user requested image:%{image}")
-    return image
-
-
-def rebuild_openstack_server(server_id, reason):
-    # Validate server_id
-    conn = openstack.connection.from_config()
-    server = conn.get_server(server_id)
 
-    image_descr = get_image_from_reason(reason)
-    if not image_descr:
-        image_descr = server.image.id
-        logger.info(
-            f"couldn't parse image from reason '%{reason}', falling back to "
-            f"existing image:%{image_descr}"
-        )
-
-    image = conn.image.find_image(image_descr)  # doesn't throw exception
-    if image is None:
-        logger.error(f"image {image_descr} either not found or not unique")
+def get_hostlist_from_slurm():
+    """
+    Parse the hostlist provided by Slurm as command-line arguments.
+    """
+    if len(sys.argv) < 2:
+        logger.error("No hostlist provided by Slurm")
         sys.exit(1)
-
-    # Note that OpenStack will power down the server as part of the rebuild
-    logger.info(f"rebuilding server %{server_id} with image %{image.id}")
-    conn.rebuild_server(server_id, image.id)
+    return sys.argv[1].split(",")
 
 
-def do_reboot():
-    # in case we are on an NFS share,
-    # try to ensure python stopped before reboot
-    os.execvp("reboot", ["reboot"])
+def read_hostvars(node):
+    """
+    Read the hostvars.yml file for a specific node.
+    """
+    hostvars_file = Path(HOSTVARS_DIR) / node / "hostvars.yml"
+    if not hostvars_file.exists():
+        logger.warning(f"No hostvars.yml found for node: {node}")
+        return None
 
+    with open(hostvars_file, "r") as f:
+        return yaml.safe_load(f)
 
-def rebuild_or_reboot():
-    """A RebootProgram for slurm which can rebuild the node running it.
 
-    This is intended to set as the `RebootProgram` in `slurm.conf`.
-    It is then triggered by slurm using something like:
-        scontrol reboot [ASAP] reason="rebuild image:<image_name_or_id>" <NODES>
+def get_current_image(conn, server_id):
+    """
+    Retrieve the current image for a server using OpenStack.
+    """
+    server = conn.get_server(server_id)
+    if not server:
+        logger.error(f"Server ID {server_id} not found in OpenStack")
+        return None
+    return server.image.id
+
+
+def rebuild_node(conn, server_id, target_image_id):
+    """
+    Rebuild the node with the target image.
+    """
+    image = conn.image.find_image(target_image_id)
```

Comment:
I think this does name or image_uuid, which is probably correct. I think you could do image.get() if its always an image uuid?